torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -0,0 +1,383 @@
+
+    // Bind the formal to the actual
+    formalToActual[formalAlias] = actualAlias;
+
+    // Record all writes
+    for (const auto& alias : actualAlias.sets()) {
+      if (formal->isWrite()) {
+        aliasToWrites_[alias].insert(node);
+      }
+    }
+  }
+
+  // Use the formal-actual mapping to give aliases to the outputs
+  for (size_t i = 0; i < schema.returns().size(); i++) {
+    const auto actual = node->outputs().at(i);
+    if (!isMutableType(actual)) {
+      continue;
+    }
+
+    const auto& formal = schema.returns()[i].alias_info();
+    if (!formal) {
+      // This is a fresh tensor
+      giveAlias(actual, getFreshAlias());
+      continue;
+    }
+
+    // We don't support composite types for alias analysis yet.
+    JIT_ASSERT(formal->containedTypes().size() == 0);
+
+    if (formal->isWildcard()) {
+      giveAlias(actual, AliasInfo::createWildcard());
+      continue;
+    }
+
+    const auto& formalAlias = formal->set();
+    auto outputAlias = formalToActual.at(formalAlias);
+
+    giveAlias(actual, outputAlias);
+  }
+}
+
+void AliasDb::analyzeIf(Node* node) {
+  // For if statements, the alias set of an output is the union of the
+  // alias sets generated by the if and else block
+  const auto trueBlock = node->blocks().at(0);
+  const auto falseBlock = node->blocks().at(1);
+  analyze(trueBlock);
+  analyze(falseBlock);
+
+  for (size_t i = 0; i < node->outputs().size(); i++) {
+    const auto nodeOutput = node->outputs()[i];
+    if (!isMutableType(nodeOutput)) {
+      continue;
+    }
+
+    const auto trueOutput = trueBlock->outputs().at(i);
+    const auto falseOutput = falseBlock->outputs().at(i);
+
+    // If a value is only used in one of the branches, then other branch
+    // won't have alias information for it. Just assign it an empty set in
+    // this case.
+    if (valueToAlias_.count(trueOutput) == 0) {
+      JIT_ASSERT(valueToAlias_.count(falseOutput) != 0);
+      giveAlias(trueOutput, AliasInfo());
+    } else if (valueToAlias_.count(falseOutput) == 0) {
+      JIT_ASSERT(valueToAlias_.count(trueOutput) != 0);
+      giveAlias(falseOutput, AliasInfo());
+    }
+
+    AliasInfo aliasInfo;
+    aliasInfo.unionWith(valueToAlias_.at(trueOutput));
+    aliasInfo.unionWith(valueToAlias_.at(falseOutput));
+    valueToAlias_[nodeOutput] = aliasInfo;
+  }
+}
+
+void AliasDb::analyzeLoop(Node* node) {
+  // For loops, we need to copy alias information from the node
+  // input/outputs to the body block input/outputs
+  const auto bodyBlock = node->blocks().at(0);
+
+  // 1. Copy alias info from node input to block input
+  const auto loopCarriedInputs = node->inputs().slice(2); // skip max, cond
+  const auto blockInputs = bodyBlock->inputs().slice(1); // skip trip
+
+  JIT_ASSERT(loopCarriedInputs.size() == blockInputs.size());
+  for (size_t i = 0; i < blockInputs.size(); i++) {
+    const auto blockInput = blockInputs[i];
+    if (!isMutableType(blockInput)) {
```

Comment:
It is useful to have `valueToAlias_` only contain mutable types, because then we can more easily detect when we are missing aliasing information (`at()` will throw). Otherwise, we may silently propagate empty alias sets.

Reply:
One way to do this in a less brittle way is to have `giveAlias` no-op if the value is non-mutable.